emcmake.py

```diff
@@ -35,7 +35,9 @@ def has_substr(args, substr):
 
   if not has_substr(args, '-DCMAKE_CROSSCOMPILING_EMULATOR'):
     node_js = config.NODE_JS[0]
-    args.append(f'-DCMAKE_CROSSCOMPILING_EMULATOR={node_js}')
+    # In order to allow cmake to run code built with pthreads we need to pass extra flags to node.
+    # See https://github.com/emscripten-core/emscripten/issues/15522
+    args.append(f'-DCMAKE_CROSSCOMPILING_EMULATOR={node_js};--experimental-wasm-threads;--experimental-wasm-bulk-memory')
```

Comment:
Note that these explicitly added options are redundant (+ could break compatibility) with newer versions of Node.js.
```bash
$ node -v
v16.13.0
$ node --v8-options | grep -A1 wasm-threads
  --experimental-wasm-threads (enable prototype thread opcodes for wasm)
        type: bool  default: true
$ node --v8-options | grep -A1 wasm-bulk-memory
$ node --experimental-wasm-bulk-memory
node: bad option: --experimental-wasm-bulk-memory
```

(You could also workaround this issue with a system-installed version of Node.js, see for example: https://github.com/emscripten-core/emscripten/issues/4848#issuecomment-932051544)

Reply:
Please correct me if I'm wrong, but I was under the assumption that we expect the users to use the version of node bundled within emsdk (`v14.15.5`); so we can update this file and the node version as a single atomic update? Alternatively, I can write a small function here to check whether each of `--experimental-wasm-threads` , `--experimental-wasm-bulk-memory` flags are available before appending them to the cmake invocation

Reply:
Sounds like we might have to check the node version before conditionally adding these flags.   That doesn't seem too unreasonable to me since this code is only run once when cmake is run

Reply:
> Please correct me if I'm wrong, but I was under the assumption that we expect the users to use the version of node bundled within emsdk (`v14.15.5`); so we can update this file and the node version as a single atomic update? Alternatively, I can write a small function here to check whether each of `--experimental-wasm-threads` , `--experimental-wasm-bulk-memory` flags are available before appending them to the cmake invocation

Sadly we cannot guarantee the exact version of node that is being used in emscripten.  We only check for the a minimum version which is currently 4.1.1.

I would write a little function here to check the version of node before adding these args.

Reply:
> Please correct me if I'm wrong, but I was under the assumption that we expect the users to use the version of node bundled within emsdk (`v14.15.5`); so we can update this file and the node version as a single atomic update?

My concern was that `emcmake` is also used outside of emsdk, so we can't ensure that users always use Node v14.5.5. For example, users installing Emscripten via [Homebrew's formula](https://formulae.brew.sh/formula/emscripten) are using [Node v17.0.1](https://github.com/Homebrew/homebrew-core/blob/47e6dc9e7fe7284066ee7159cace7201e7e315fb/Formula/node.rb#L4) (at the time of writing).

Reply:
Couple of updates -- it looks like `--experimental-wasm-bulk-memory` is redundant for the case I am testing here, so I can drop it entirely; So then we just need to add `--experimental-wasm-threads` for now which is supported by all node versions released so far. To be future proof, I can add a check for whether `--experimental-wasm-threads` is accepted by the node binary (by executing `node --experimental-wasm-threads`) , or we can just leave it here until node removes this flag and then add a version check. WDYT @sbc100 ?

Reply:
Why is `--experimental-wasm-bulk-memory` redundant?  My understanding is the our pthread support requires bulk memory to be enabled.   Is node automatically enabling bulk memory when thread support is enabled?  Or maybe your test cases so small that they don't end up using any bulk memory instruction.    For pthread-based code we generate bulk memory instruction to load static data into memory in the `__wasm_init_memory` function that the linker generates.   Unless you have no static data at all I think will need this.

Reply:
For the issue I reported, `check_type_size` in CMake is failing with pthreads enabled, and setting `--experimental-wasm-threads` seems to be sufficient. Looking at the template for checking the type size, it does seem to contain some globals (https://github.com/thewtex/emscripten/blob/b206b9e304749a2132e5da0d0aa594b92151daef/cmake/Modules/CheckTypeSize.c.in) but I'm not sure if that requires bulk memory support

Reply:
Maybe we should add `--experimental-wasm-bulk-memory` in a separate PR with an appropriate test case which covers this case (I'd have to spend some additional time figuring out how to ensure bulk memory load instructions are exercised in the test)?

Reply:
> Why is `--experimental-wasm-bulk-memory` redundant?

It seems that it is already enabled by default in the bundled Node (see https://github.com/v8/v8/commit/af1988f1e43d663167ea1566983aa3a7ff625701).
```bash
emsdk/node/14.15.5_64bit/bin/node --experimental-wasm-bulk-memory --v8-options | grep -A1 wasm-bulk-memory
  --experimental-wasm-bulk-memory (enable prototype bulk memory opcodes for wasm)
        type: bool  default: true
```

Reply:
Ah, that explains it; since `--experimental-wasm-threads;--experimental-wasm-bulk-memory=false` also results in a failure for my test. What's the oldest node version we need to support @sbc100 ? I can see if `--experimental-wasm-bulk-memory` was enabled by default for that version

Reply:
The minimum version of node that we support is 4.1.1:

https://github.com/emscripten-core/emscripten/blob/738a94050bab855497b20ea6c530ac6634e6a7fd/tools/shared.py#L49

Reply:
I'm also happy to land with without `--experimental-wasm-bulk-memory`.    this is kind of niche case and can't imagine many folks who need to to work and happen to be such on an old node version.

Reply:
Removed the `--experimental-wasm-bulk-memory` flag and added a comment; Not adding the node version check for now for `--experimental-wasm-threads` since it is present even in latest node version